src/rideables/TGraph.hpp

```diff
@@ -161,171 +258,240 @@ class TGraph : public RGraph{
+                lock(vid);
+                if (idxToVertex[vid] == nullptr) {
+                    unlock(vid);
+                    return false;
+                }
+                uint32_t seq = get_seq(vid);
+                for (auto r : source(vid)) {
                     vertices.push_back(r->dest);
                 }
-                for (Relation *r : v->dest_list) {
+                for (auto r : destination(vid)) {
                     vertices.push_back(r->src);
                 }
                 
-                vertices.push_back(id);
+                vertices.push_back(vid);
                 std::sort(vertices.begin(), vertices.end()); 
                 vertices.erase(std::unique(vertices.begin(), vertices.end()), vertices.end());
 
                 // Step 2: Release lock, then acquire lock-order...
-                v->unlock();
+                unlock(vid);
                 for (int _vid : vertices) {
-                    idxToVertex[_vid]->lock();
+                    lock(_vid);
+                    if (idxToVertex[_vid] == nullptr && get_seq(vid) == seq) {
+                        for (auto r : source(vid)) {
+                            if (r->dest == _vid)
+                            std::cout << "(" << r->src << "," << r->dest << ")" << std::endl;
+                        }
+                        for (auto r : destination(vid)) {
+                            if (r->src == _vid)
+                            std::cout << "(" << r->src << "," << r->dest << ")" << std::endl;
+                        }
+                        std::abort();
+                    }
                 }
 
                 // Has vertex been changed? Start over
-                if (v->seqNumber != seq) {
+                if (get_seq(vid) != seq) {
                     std::reverse(vertices.begin(), vertices.end());
                     for (int _vid : vertices) {
-                        idxToVertex[_vid]->unlock();
+                        unlock(_vid);
                     }
                     goto startOver;
                 }
 
                 // Has not changed, continue...
                 // Step 3: Remove edges from all other
                 // vertices that relate to this vertex
-                std::vector<Relation*> garbageList;
-                for (int _vid : vertices) {
-                    if (_vid == id) continue;
-                    Vertex *_v = idxToVertex[_vid];
-                    std::vector<Relation*> toRemoveList;
+                for (int other : vertices) {
+                    if (other == vid) continue;
 
-                    for (Relation *r : _v->adjacency_list) {
-                        if (r->src == id) {
-                            toRemoveList.push_back(r);
+                    Relation src(other, vid, -1);
+                    Relation dest(vid, other, -1);
+                    if (!has_relation(source(other), &src) && !has_relation(destination(other), &dest)) {
```

Comment:
My fault for not catching this: It just has to find that _one_ of these relations exist because we extract vertices from both `source` and `destination` sets, it isn't necessary for it to be in both. We don't keep track of which set the vertex `other` belongs to, just that it was an edge in `source` or `destination`.